netket/nn/utils.py

```diff
@@ -144,3 +147,55 @@ def fix_one_kernel(args):
         return (path, array)
 
     return unflatten_dict(dict(map(fix_one_kernel, flatten_dict(params).items())))
+
+
+def _get_output_idx(
+    shape: Tuple[int, ...], max_bits: Optional[int] = None
+) -> Tuple[Tuple[int, ...], int]:
+    bits_per_local_occupation = tuple(np.ceil(np.log2(shape)).astype(int))
+    if max_bits is None:
+        max_bits = max(bits_per_local_occupation)
+    output_idx = []
+    offset = 0
+    for b in bits_per_local_occupation:
+        output_idx.extend([i + offset for i in range(b)][::-1])
+        offset += max_bits
+    output_idx = tuple(output_idx)
+    return output_idx, max_bits
+
+
+@partial(jax.jit, static_argnums=(0, 2, 3))
+def binary_encoding(
+    shape: Union[DiscreteHilbert, Tuple[int, ...]],
+    x: Array,
+    *,
+    max_bits: Optional[int] = None,
+) -> Array:
+    """
+    Encodes the array `x` into a set of binary-encoded variables described by shape.
+    The i-th element of x will be encoded in ceil(log2(shape[i])) bits.
+
+    Args:
+        shape: Either a discrete Hilbert space or a tuple containing the shape of the
+            discrete Hilbert space.
+        x: The array to encode.
+        max_bits: The maximum number of bits to use for each element of `x`.
+    """
+    if isinstance(shape, DiscreteHilbert):
+        shape = tuple(shape.shape)
```

Comment:
@VolodyaCO if you rebase this branch on master you'll be able to use this hilbert.sstates_to_local_indices which does exactly what you need here, for any Hilbert space (even tensor).

```suggestion
    if isinstance(shape, DiscreteHilbert):
        x = shape.states_to_local_indices(x)
        shape = tuple(shape.shape)
```